### PR TITLE
Introducing HTML parser/renderer.

### PR DESCRIPTION
**Pull request by MarcinSc**

### CODE REVIEW

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/170/
Hooray Jenkins reported success with all tests good!

**Comment by immortius** [comment]
Without having reviewed this in detail can I suggest moving all of browser related classes into a single package, and having packages grouping by fuction under that. This would help keep all the related classes together rather than jumbled with all the other nui types.

**Comment by immortius** [comment]
Or if it is all under widgets.browser I would suggest pulling it out of widgets?

**Comment by immortius** [comment]
On further thought, ignore all that. Trying to review code on my phone was throwing me off.

**Comment by MarcinSc** [comment]
Any further comments on this?

**Comment by Cervator** [comment]
@immortius @Josharias - ping!

File: engine-tests/src/test/java/org/terasology/rendering/nui/widgets/browser/data/html/HTMLParserTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.nui.widgets.browser.data.html;
+
+import org.junit.Test;
+import org.terasology.rendering.nui.widgets.browser.data.DocumentData;
+import org.terasology.rendering.nui.widgets.browser.data.ParagraphData;
+import org.terasology.rendering.nui.widgets.browser.data.basic.FlowParagraphData;
+import org.xml.sax.SAXException;
```
Comment (immortius) [inline engine-tests/src/test/java/org/terasology/rendering/nui/widgets/browser/data/html/HTMLParserTest.java]: Doesn't look like SAXException is thrown by your code, you can probably drop this from the test.

File: engine/src/main/java/org/terasology/rendering/nui/widgets/browser/data/basic/FlowParagraphRenderable.java [hunk 2]
```diff
@@ -115,56 +84,27 @@ public void render(Canvas canvas, Rect2i region, TextRenderStyle defaultStyle, H
...
-            this.insetWidth = insetWidth;
-            this.insetHeight = insetHeight;
-            this.laidLines = laidLines;
-        }
+    private Iterable<LaidFlowLine<FlowRenderable>> updateCacheIfNeeded(ParagraphRenderStyle defaultStyle, int yStart, ContainerRenderSpace containerRenderSpace) {
+        // TODO introduce cache, once the RenderSpace gets stabilized and allows comparing
+        return FlowLineBuilder.getLines(flowParagraphData, defaultStyle, yStart, containerRenderSpace);
     }
+//
+//    private static final class LineBuilderCache {
```
Comment (immortius) [inline engine/src/main/java/org/terasology/rendering/nui/widgets/browser/data/basic/FlowParagraphRenderable.java]: Might want to clean this up. *shrug*

**Comment by immortius** [comment]
Generally looks good.

It would be nice to have a bit of javadoc, at least on the classes and packages explaining their purpose.

The little thing that catches my eye is the ContainerInteger family of classes. I can see what they are for, but think the name could be clearer - perhaps SizeHint or SizeAttribute? Something that reflects its role better.

**Comment by MarcinSc** [comment]
I'll keep on working on it, the layout doesn't work the way it works in browsers (I want to get as close to it as possible), and I'll be adding support for borders, so I might do the suggested changes in the next code drop, once I get some time to work on it.

For now, it is in a usable state, so I hope it's ok if we merge it "as is".

**Comment by immortius** [comment]
Fair enough. I'll hopefully have something for you to integrate soon too.

**Comment by msteiger** [comment]
Where can I find documentation on this? Which tags are supported? How can I embed an image?

Maybe this info could live in a Wiki page?

**Pull request merged**